server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -70,30 +81,104 @@ private ImmutableMap<String, ProgramDefinition> mapNameToProgram(
     return mapNameToProgramWithFilter(repository, service, versionModel, Optional.empty());
   }
 
-  private ActiveAndDraftPrograms(VersionRepository repository, Optional<ProgramService> service) {
+  private ActiveAndDraftPrograms(
+      VersionRepository repository,
+      Optional<ProgramService> service,
+      EnumSet<ActiveAndDraftProgramsType> types) {
     VersionModel active = repository.getActiveVersion();
     VersionModel draft = repository.getDraftVersionOrCreate();
     // Note: Building this lookup has N+1 query behavior since a call to getProgramDefinition does
     // an additional database lookup in order to sync the set of questions associated with the
     // program.
+
+    // Active, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> activeNameToProgram =
+        mapNameToProgramWithFilter(repository, service, active, Optional.of(DisplayMode.DISABLED));
+
+    // All active programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> activeNameToProgramAll =
         mapNameToProgram(repository, service, active);
 
+    // Draft, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> draftNameToProgram =
+        mapNameToProgramWithFilter(repository, service, draft, Optional.of(DisplayMode.DISABLED));
+
+    // All draft programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> draftNameToProgramAll =
         mapNameToProgram(repository, service, draft);
 
-    this.activePrograms = activeNameToProgram.values().asList();
-    this.draftPrograms = draftNameToProgram.values().asList();
-    this.versionedByName =
-        Sets.union(activeNameToProgram.keySet(), draftNameToProgram.keySet()).stream()
-            .collect(
-                ImmutableMap.toImmutableMap(
-                    Function.identity(),
-                    programName -> {
-                      return Pair.create(
-                          Optional.ofNullable(activeNameToProgram.get(programName)),
-                          Optional.ofNullable(draftNameToProgram.get(programName)));
-                    }));
+    this.activePrograms =
```

Comment:
**suggestion:** I think this structure may be cleaner since we already have to go into an if condition, but let me know what you think:

```
if (types.containsAll(allProgramTypes)) {
this.activePrograms = ..
this.draftPrograms = ..
} else {
this.activePrograms = ..
this.draftPrograms = ..

  if (types.contains(ActiveAndDraftProgramsType.DISABLED)) {
  ...
  }

  if (types.contains(...IN_USE
  ...
}
```

Reply:
One problem with this structure is that with `if (types.contains(...IN_USE`, it will cause error `variable versionedByName might already have been assigned`
So I changed it to `else if (types.contains(...IN_USE`. But the rest change you suggested has been applied.

Reply:
Sounds good, thanks!!